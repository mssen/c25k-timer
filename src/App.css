.main {
  margin: 20px;
  font-family: Arial, Helvetica, sans-serif;
  display: flex;
  flex-direction: column;
}

.center {
  align-self: center;
}

.horizontal {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
}

.main-header {
  margin: 0;
  font-size: 50px;
}

.sub-header {
  color: gray;
  text-transform: uppercase;
}

.interval-timer {
  margin: 40px 0;
  display: flex;
  align-items: baseline;
}

.interval-text {
  text-transform: uppercase;
  font-size: 40px;
}

.play-button {
  cursor: pointer;
  border-radius: 4px;
  font-size: 50px;
  background: white;
  padding: 10px 20px 10px 20px;
  border: 5px solid black;
}

.gray {
  color: gray;
}