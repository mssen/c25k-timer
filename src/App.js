import React, { Component } from 'react';

import Timer from './components/Timer'
import './App.css';

class App extends Component {
  render() {
    return (
      <div className="main">
        <Timer />
      </div>
    );
  }
}

export default App;
