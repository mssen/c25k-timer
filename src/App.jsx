import React from 'react';

import Timer from './components/Timer';
import Navigation from './components/Navigation';
import IntervalTracker from './components/IntervalTracker';
import PickerModal from './components/PickerModal';

import './App.css';
import { SEGMENTS } from './data/segments';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      segment: 0,
      interval: 0,
      showModal: false
    };

    this.updateSegment = this.updateSegment.bind(this);
    this.updateInterval = this.updateInterval.bind(this);
    this.hideModal = this.hideModal.bind(this);
    this.showModal = this.showModal.bind(this);
  }

  updateSegment(adjustment) {
    this.setState((prevState) => ({
      segment: prevState.segment + adjustment,
      interval: 0
    }));
  }

  updateInterval(newInterval) {
    this.setState({ interval: newInterval });
  }

  showModal() {
    this.setState({
      showModal: true
    })
  }

  hideModal() {
    this.setState({
      showModal: false
    })
  }

  getIntervalType() {
    const currentInterval = this.state.interval;
    if (currentInterval === 0) {
      return 'Warmup';
    } else if (currentInterval === SEGMENTS[this.state.segment].intervals.length - 1) {
      return 'Cooldown';
    } else if (currentInterval % 2 === 1) {
      return 'Run';
    } else {
      return 'Walk';
    }
  }

  render() {
    const segment = SEGMENTS[this.state.segment];

    return (
      <div className="main">
        <h1>Week { segment.week }</h1>
        <h1>Day { segment.day }</h1>

        <div>
          { this.getIntervalType() }
          <Timer
            intervals={ segment.intervals }
            updateInterval={ this.updateInterval }
            segment={ this.state.segment } />
        </div>

        <div>
          Total Time
          <Timer
            intervals={ [ segment.intervals.reduce((prev, curr) => prev + curr) ] }
            updateInterval={ () => '' }
            segment={ this.state.segment } />
        </div>

        <IntervalTracker current={ this.state.interval } total={ segment.intervals.length - 2 } />

        <Navigation
          week={ segment.week }
          day={ segment.day }
          triggerNavigation={ this.updateSegment }
          showModal={ this.showModal } />

        {
          this.state.showModal
            && <PickerModal segments={ SEGMENTS }
                currentSegment={ this.state.segment }
                updateSegment={ this.updateSegment }
                hideModal={ this.hideModal } />
        }
      </div>
    );
  }
}

export default App;
