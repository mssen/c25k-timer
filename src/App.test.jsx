import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';

jest.useFakeTimers();

it('renders without crashing', () => {
  shallow(<App />);
});

it('can start/pause', () => {
  const wrapper = mount(<App />);
  const playPauseButton = wrapper.find('button').first();
  expect(wrapper.find('Timer')).toMatchSnapshot();

  // Play
  playPauseButton.simulate('click');
  jest.runTimersToTime(1000);
  expect(wrapper.find('Timer')).toMatchSnapshot();

  // Pause
  playPauseButton.simulate('click');
  jest.runTimersToTime(1000);
  expect(wrapper.find('Timer')).toMatchSnapshot();
});

it('hide/shows the modal', () => {
  const wrapper = mount(<App />);
  expect(wrapper.find('PickerModal').exists()).toBe(false);

  // Show
  wrapper.setState({ showModal: true });
  expect(wrapper.find('PickerModal').exists()).toBe(true);

  // Hide
  wrapper.setState({ showModal: false });
  expect(wrapper.find('PickerModal').exists()).toBe(false);
});

describe('gets correct interval type', () => {
  it('gets warmup', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ interval: 0 });
    expect(wrapper.find('div > div').first()).toMatchSnapshot();
  });

  it('gets cooldown', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ interval: 3 });
    expect(wrapper.find('div > div').first()).toMatchSnapshot();
  });

  it('gets run', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ interval: 1 });
    expect(wrapper.find('div > div').first()).toMatchSnapshot();
  });

  it('gets walk', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ interval: 2 });
    expect(wrapper.find('div > div').first()).toMatchSnapshot();
  });
});