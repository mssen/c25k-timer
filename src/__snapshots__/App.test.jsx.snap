// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can start/pause 1`] = `
Array [
  <Timer
    intervals={
        Array [
            2,
            5,
            10,
            2,
          ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
>
    <div>
        0:00 / 0:02
    </div>
</Timer>,
  <Timer
    intervals={
        Array [
            19,
          ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
>
    <div>
        0:00 / 0:19
    </div>
</Timer>,
]
`;

exports[`can start/pause 2`] = `
Array [
  <Timer
    intervals={
        Array [
            2,
            5,
            10,
            2,
          ]
    }
    play={true}
    segment={0}
    updateInterval={[Function]}
>
    <div>
        0:01 / 0:02
    </div>
</Timer>,
  <Timer
    intervals={
        Array [
            19,
          ]
    }
    play={true}
    segment={0}
    updateInterval={[Function]}
>
    <div>
        0:01 / 0:19
    </div>
</Timer>,
]
`;

exports[`can start/pause 3`] = `
Array [
  <Timer
    intervals={
        Array [
            2,
            5,
            10,
            2,
          ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
>
    <div>
        0:01 / 0:02
    </div>
</Timer>,
  <Timer
    intervals={
        Array [
            19,
          ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
>
    <div>
        0:01 / 0:19
    </div>
</Timer>,
]
`;

exports[`gets correct interval type gets cooldown 1`] = `
<div>
  Cooldown
  <Timer
    intervals={
      Array [
        2,
        5,
        10,
        2,
      ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
  />
</div>
`;

exports[`gets correct interval type gets run 1`] = `
<div>
  Run
  <Timer
    intervals={
      Array [
        2,
        5,
        10,
        2,
      ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
  />
</div>
`;

exports[`gets correct interval type gets walk 1`] = `
<div>
  Walk
  <Timer
    intervals={
      Array [
        2,
        5,
        10,
        2,
      ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
  />
</div>
`;

exports[`gets correct interval type gets warmup 1`] = `
<div>
  Warmup
  <Timer
    intervals={
      Array [
        2,
        5,
        10,
        2,
      ]
    }
    play={false}
    segment={0}
    updateInterval={[Function]}
  />
</div>
`;
